PE-22051 Changes for adding loadbalancer to LEI job

shaigy opened the pull request:
Changes for enabling installation of agent nodes separately from
other PE nodes so that agents can be installed pointing at loadbalancer.
Also modified deploy_frictionless_to_master() to add the pe_repo
packages to ‘PE Master’ group instead of 'Beaker Frictionless
Agent’group
Made following changes:
a) Added methods for filtering out agent nodes with 'lb_connect' role
and to check if host/loadbalancer has 'lb_connect' role
b) Added a method for agent only installation
c) If there are laodbalancer and agent nodes with role 'lb_connect',
use loadbalancer as downloadhost for frictionless install on those nodes
d) Add pe_repo packages to PE Master group instead of old 'Beaker
Frictionless Agent’ group
e) After downloading the agent tar balls on master, run ‘puppet agent
-t’ on compile_masters to populate them as well

Please delete any headings that don't apply to this Pull Request (PR).

#### What's this PR do?
#### Who would you like to review this PR?

@puppetlabs/integration, @puppetlabs/beaker (repo owners)

#### Should any of this be tested outside the normal PR CI cycle?
#### Any background context you want to provide?
#### Questions for reviewers?

puppetlabs-jenkins commented:
Can one of the admins verify this patch?

shaigy commented:
@ericwilliamson Modified the PR with changes we talked about. Can you please review?

shaigy commented:
Validated using following configuration that failed PEZ before:
---
HOSTS:
  sles12-64-1:
    pe_dir: http://enterprise.delivery.puppetlabs.net/2018.1/ci-ready/
    pe_ver: 2018.1.0-rc5
    pe_upgrade_dir:
    pe_upgrade_ver:
    hypervisor: vmpooler
    platform: sles-12-x86_64
    packaging_platform: sles-12-x86_64
    template: sles-12-x86_64
    roles:
    - agent
    - master
    - dashboard
    - database
  sles12-64-2:
    pe_dir: http://enterprise.delivery.puppetlabs.net/2018.1/ci-ready/
    pe_ver: 2018.1.0-rc5
    pe_upgrade_dir:
    pe_upgrade_ver:
    hypervisor: vmpooler
    platform: sles-12-x86_64
    packaging_platform: sles-12-x86_64
    template: sles-12-x86_64
    roles:
    - agent
    - frictionless
    - compile_master
    main:
      dns_alt_names: puppet
      environmentpath: "/etc/puppetlabs/puppet/environments"
  osx1012-64-1:
    pe_dir: http://enterprise.delivery.puppetlabs.net/2018.1/ci-ready/
    pe_ver: 2018.1.0-rc5
    pe_upgrade_dir:
    pe_upgrade_ver:
    hypervisor: vmpooler
    platform: osx-10.12-x86_64
    packaging_platform: osx-10.12-x86_64
    template: osx-1012-x86_64
    roles:
    - agent
    - frictionless
CONFIG:
  nfs_server: none
  consoleport: 443
  pooling_api: http://vmpooler.delivery.puppetlabs.net/

$ bundle exec beaker -h ~/beaker/pez.conf --helper lib/beaker_helper.rb --pre-suite setup/scale/install

              - Test Case Summary for suite 'pre_suite' -
       Total Suite Time: 1070.89 seconds
      Average Test Time: 62.99 seconds
              Attempted: 17
                 Passed: 10
                 Failed: 0
                Errored: 0
                Skipped: 7
                Pending: 0
                  Total: 17

      - Specific Test Case Status -

Failed Tests Cases:
Errored Tests Cases:
Skipped Tests Cases:
  Test Case setup/scale/install/00_pre_install/20_rpm_setup.rb
  Test Case setup/scale/install/20_classification/05_loadbalancer.rb
  Test Case setup/scale/install/20_classification/30_clamps.rb
  Test Case setup/scale/install/20_classification/40_metrics.rb
  Test Case setup/scale/install/20_classification/45_openldap.rb
  Test Case setup/scale/install/20_classification/50_ntp.rb
  Test Case setup/scale/install/20_classification/60_loadbalance_agents.rb
Pending Tests Cases:

              No tests to run for suite 'tests'
              No tests to run for suite 'post_suite'
              No tests to run for suite 'pre_cleanup'
              Cleanup: cleaning up after successful run
              Handing 'd1qu527eu5vpw25.delivery.puppetlabs.net' back to vmpooler for VM destruction
              Handing 'ak9wdm0wmuys7rd.delivery.puppetlabs.net' back to vmpooler for VM destruction
              Handing 'qmdax0o5tr2gjq4.delivery.puppetlabs.net' back to vmpooler for VM destruction
              Spent 0.44 seconds cleaning up
Beaker completed successfully, thanks.

shaigy commented on lib/beaker-pe/install/pe_utils.rb:
```diff
@@ -101,6 +101,30 @@ def install_ca_cert_on(host, opts)
+
+        # Returns true if loadbalncer exists and is configured with 'lb_connect' role
+        def lb_connect_loadbalancer_exists?
+          if (any_hosts_as?('loadbalancer'))
```
Will do

shaigy commented on lib/beaker-pe/install/pe_utils.rb:
```diff
@@ -124,6 +148,14 @@ def frictionless_agent_installer_cmd(host, opts, pe_version)
 
+          # If this is an agent node configured to connect to the loadbalancer
+          # using 'lb_connect' role, then use loadbalancer in the download url
+          # instead of master (PE-22051)
```
Sure

shaigy commented on lib/beaker-pe/install/pe_utils.rb:
```diff
@@ -376,33 +408,43 @@ def deploy_frictionless_to_master(host)
+            # compilemasters(PE-22051)
+            # Previosuly, we were creating a new node group 'Beaker Frictionless Agent' and adding
+            # pe_repo packages to that group
+            #
```
I also had second thoughts here and you confirmed it. Will remove that comment except the first line

shaigy commented on lib/beaker-pe/install/pe_utils.rb:
```diff
@@ -1549,6 +1591,34 @@ def download_pe_conf_if_master(host)
+        # @param [Array] agent only nodes from Beaker hosts
+        # @param [Hash] opts The Beaker options hash
+        #
+        def install_agents_only_on(agentnodes, opts)
```
I am inclined to add that change as a different PR after this goes in since that will be a change affecting almost all of the CI jobs and if there is a failure it will be easy to trace back. What do you think?

shaigy replied:
As we discussed over chat, I have added these changes. Now simple_monolithic_install() will call install_agents_only_on() for installing agents.

shaigy commented on spec/beaker-pe/install/pe_utils_spec.rb:
```diff
@@ -188,6 +199,22 @@ def logger
+  end
+
+  describe 'get_lb_downloadhost' do
+    it 'choose lb_connect loadbalancer, if there is one' do
```
Good point. I will add one

shaigy commented on spec/beaker-pe/install/pe_utils_spec.rb:
```diff
@@ -1458,6 +1503,22 @@ def slice_installer_options(host)
     end
 
+  describe 'install_agents_only_on' do
+    let(:lb_agent) { make_host('agent', :pe_ver => '2016.4', :platform => 'debian-7-x86_64', :roles => ['frictionless', 'lb_connect']) }
```
Nice, will do

shaigy commented on spec/beaker-pe/install/pe_utils_spec.rb:
```diff
@@ -1458,6 +1503,22 @@ def slice_installer_options(host)
+
+      allow(subject).to receive(:stop_agent_on)
+      allow(subject).to receive(:sign_certificate_for)
+    end
```
No. Just a single check to see agent is getting installed on an agent host
```
allow(subject).to receive(:installer_cmd).with(lb_agent, anything()).and_return("install lb agent")
```

shaigy replied:
Added new tests

puppetlabs-jenkins commented:
Build finished.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker-pe_btc-intn/251/

shaigy commented:
@ericwilliamson can you please review the changes? All of the requested changes are in except the refactoring of simple_monolithic_install. I am inclined to add that as a separate PR once this change is in, since that change will affect almost all of the CI jobs.

puppetlabs-jenkins commented:
Build finished.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker-pe_btc-intn/252/

ericwilliamson commented on lib/beaker-pe/install/pe_utils.rb:
```diff
@@ -101,6 +101,30 @@ def install_ca_cert_on(host, opts)
+        #@param [Host] agent host with lb_connect role
+        def get_lb_downloadhost(host)
+          downloadhost = master
+          if( ! host['roles'].include?('loadbalancer') &&  lb_connect_loadbalancer_exists?)
```
@shaigy Sorry for not making it clear - the style concerns applied to the entire PR I just commented on the first one.

Can you please go through and remove parens from if statements etc.

shaigy replied:
Done! Hopefully I have covered everything this time.

shaigy commented:
@ericwilliamson all the requested changes are in now.

puppetlabs-jenkins commented:
Build finished.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker-pe_btc-intn/254/

ericwilliamson commented on lib/beaker-pe/install/pe_utils.rb:
```diff
@@ -1549,6 +1555,34 @@ def download_pe_conf_if_master(host)
+        # including infrastructure nodes, loadbalancer, hubs and spokes
+        # @param [Array] agent only nodes from Beaker hosts
+        # @param [Hash] opts The Beaker options hash
+        def install_agents_only_on(agent_nodes, opts)
```
@shaigy Is there a reason you didn't directly copy paste? I would prefer to keep the step blocks around.

shaigy replied:
Ah, ok. Will update it shortly

ericwilliamson replied:
Thanks, thats the last change I have, then this and the beaker-pe-lei one is good to go if you've re-ran the tests and verified it all works still.

shaigy replied:
Done! Validated with couple of configurations. I will add a comment with test result.

shaigy commented:
@ericwilliamson All of the requested changes are in.
Validated with following configurations:
1) mono install and split lei upgrade with 'lb_connect' role for loadbalancer and agent
2) mono upgrade and split install with LEI configuration used in CI
3) frictionless agent install/upgrade with mono master
Also ran acceptance tests against most of them
```
$ bundle exec beaker --pre-suite setup/scale/install/,setup/scale/upgrade --tests acceptance/tests/ --helper lib/beaker_helper.rb -h mono-lei.cfg
             - Test Case Summary for suite 'tests' -
       Total Suite Time: 1657.04 seconds
      Average Test Time: 28.57 seconds
              Attempted: 58
                 Passed: 50
                 Failed: 0
                Errored: 0
                Skipped: 8
                Pending: 0
                  Total: 58

      - Specific Test Case Status -

```

puppetlabs-jenkins commented:
Build finished.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker-pe_btc-intn/255/
